exist-core/src/main/java/org/exist/xquery/WindowExpr.java

```diff
@@ -0,0 +1,684 @@
+
+        // bind the value
+        windowReturnVariable.setValue(windowContextVariable.getValue());
+
+        // declare it
+        context.declareVariableBinding(windowReturnVariable);
+
+        return windowReturnVariable;
+    }
+
+    private static class WindowContextVariables {
+        final @Nullable
+        LocalVariable currentItem;
+        final @Nullable
+        LocalVariable posVar;
+        final @Nullable
+        LocalVariable previousItem;
+        final @Nullable
+        LocalVariable nextItem;
+
+        private WindowContextVariables(@Nullable final LocalVariable currentItem, @Nullable final LocalVariable posVar, @Nullable final LocalVariable previousItem, @Nullable final LocalVariable nextItem) {
+            this.currentItem = currentItem;
+            this.posVar = posVar;
+            this.previousItem = previousItem;
+            this.nextItem = nextItem;
+        }
+
+        public boolean isDefined() {
+            return currentItem != null || posVar != null || previousItem != null || nextItem != null;
+        }
+
+        public void destroy(final XQueryContext context, final Sequence contextSequence) {
+            if (nextItem != null) {
+                nextItem.destroy(context, contextSequence);
+            }
+            if (previousItem != null) {
+                previousItem.destroy(context, contextSequence);
+            }
+            if (posVar != null) {
+                posVar.destroy(context, contextSequence);
+            }
+            if (currentItem != null) {
+                currentItem.destroy(context, contextSequence);
+            }
+        }
+    }
+
+    @Override
+    public Set<QName> getTupleStreamVariables() {
+        final Set<QName> vars = new HashSet<>();
+        if (varName != null) {
+            vars.add(varName);
+        }
+
+        if (windowStartCondition.getPreviousItem() != null) {
+            vars.add(windowStartCondition.getPreviousItem());
+        }
+        if (windowStartCondition.getCurrentItem() != null) {
+            vars.add(windowStartCondition.getCurrentItem());
+        }
+        if (windowStartCondition.getNextItem() != null) {
+            vars.add(windowStartCondition.getNextItem());
+        }
+
+        if (windowEndCondition != null) {
+            if (windowEndCondition.getPreviousItem() != null) {
+                vars.add(windowEndCondition.getPreviousItem());
+            }
+            if (windowEndCondition.getCurrentItem() != null) {
+                vars.add(windowEndCondition.getCurrentItem());
+            }
+            if (windowEndCondition.getNextItem() != null) {
+                vars.add(windowEndCondition.getNextItem());
+            }
+        }
+
+        final QName var = getVariable();
+        if (var != null) {
+            vars.add(var);
+        }
+
+        final LocalVariable startVar = getStartVariable();
+        if (startVar != null) {
+            vars.add(startVar.getQName());
+        }
+
```

Comment:
Maby extracting the check for null and add into a simple function would make this method more compact with something like this:
```
    @Override
    public Set<QName> getTupleStreamVariables() {
        final Set<QName> vars = new HashSet<>();
        addIfNotNull(vars, varName);
        addIfNotNull(vars, windowStartCondition.getPreviousItem());
        addIfNotNull(vars, windowStartCondition.getCurrentItem());
        addIfNotNull(vars, windowStartCondition.getNextItem());

        if (windowEndCondition != null) {
           addIfNotNull(vars, windowEndCondition.getPreviousItem());
           addIfNotNull(vars, windowEndCondition.getCurrentItem());
           addIfNotNull(vars, windowEndCondition.getNextItem());
        }

        addIfNotNull(vars, getVariable());

        final LocalVariable startVar = getStartVariable();
        if (startVar != null) {
            vars.add(startVar.getQName());
        }

        return vars;
    }

    private void addIfNotNull(Set<QName> vars, String value) {
      if (value != null) {
        vars.add(value);
      }
    }
```

Reply:
Done.